File: 0988-smallest-string-starting-from-leaf/0988-smallest-string-starting-from-leaf.py
class Solution:
    def smallestFromLeaf(self, root: Optional[TreeNode]) -> str:
        def get(node, route):
            route.append(chr(node.val + ord('a')))
            if not node.left and not node.right:
                if result[0] is None:
                    result[0] = ''.join(route[::-1])
                else:
                    result[0] = min(result[0], ''.join(route[::-1]))
            if node.left:
                get(node.left, route)
            if node.right:
                get(node.right, route)
            route.pop()


        result = [None]
        get(root, list())
        return result[0]

File: 0988-smallest-string-starting-from-leaf/0988-smallest-string-starting-from-leaf.java
class Solution {
    private String result;

    private void get(TreeNode node, StringBuilder route) {
        route.insert(0, (char)('a' + node.val));
        if (node.left == null && node.right == null) {
            String target = route.toString();
            if (result == null || target.compareTo(result) < 0) {
                result = target;
            }
        }
        if (node.left != null) {
            get(node.left, route);
        }
        if (node.right != null) {
            get(node.right, route);
        }
        route.deleteCharAt(0);
    }

    public String smallestFromLeaf(TreeNode root) {
        result = null;
        get(root, new StringBuilder());
        return result;
    }
}